OptimizerArgs

rasbt opened the pull request:
This PR unbundles the OptimizerArgs approach from GaLore in #1192.


## Todos

- [x] OptimizerArgs for full finetuning
  - [x] Update code 
  - [x] Add docstrings
  - [x] Update docs
  - [x] Update config files
 
- [x] OptimizerArgs for LoRA
  - [x] Update code 
  - [x] Add docstrings
  - [x] Update docs
  - [x] Update config files

- [x] OptimizerArgs for Adapter
  - [x] Update code 
  - [x] Add docstrings
  - [x] Update docs
  - [x] Update config files

- [x] OptimizerArgs for Adapter v2
  - [x] Update code 
  - [x] Add docstrings
  - [x] Update docs
  - [x] Update config files

- [x] OptimizerArgs for Pretraining
  - [x] Update code 
  - [x] Add docstrings
  - [x] Update docs
  - [x] Update config files
  - [ ] add back fusing


- [x] ensure that both `--optimizer torch.optim.AdamW` and `--optimizer AdamW` works
- [x] Add tests

rasbt commented:
Your jsonargparse example has been super helpful for understanding things a bit more @carmocca . Many thanks for this!

 But maybe it's because it's Fri evening but my brain is just not working today. I've just been banging my head against how I would get this into the finetuning method's `setup`. 

Adding the `optimizer` subclass to the parser and then calling the finetune command yields a 

```
  File "/home/zeus/miniconda3/envs/cloudspace/bin/litgpt", line 8, in <module>
    sys.exit(main())
  File "/teamspace/studios/this_studio/litgpt/litgpt/__main__.py", line 145, in main
    fn(**kwargs)
TypeError: setup() got an unexpected keyword argument 'optimizer.class_path'
```

But we can't add `optimizer` argument to the  finetuning `setup` signature because it's a duplicate command then. Conceptually, I am kind of stuck here.

Also, how would we get the `args` in 

```
optimizer = instantiate_class(model.parameters(), init=args["optimizer"])
```

if we don't pass them on from the `main()` function in `__main.py__`. I suppose if I would do the

```
parser = jsonargparse.ArgumentParser()
parser.add_subclass_arguments(torch.optim.Optimizer, "optimizer", instantiate=False, fail_untyped=False, skip={"params"})
args = parser.parse_args()
```

in the finetuning script, but then it would erase all the previous arguments.

carmocca commented on litgpt/utils.py:
```diff
@@ -486,3 +486,23 @@ def choose_logger(
     raise ValueError(f"`--logger_name={logger_name}` is not a valid option. Choose from 'csv', 'tensorboard', 'wandb'.")
+
+
+def instantiate_class(args: Union[Any, Tuple[Any, ...]], init: Dict[str, Any]) -> Any:
```
You can also import it from `lightning.pytorch.cli`

carmocca commented:
As far as integrating into  the scripts, I would:

Create an optimizer argument in https://github.com/Lightning-AI/litgpt/blob/36c6a77435d75872f525848ee1570467d120ae80/litgpt/finetune/lora.py#L40

To avoid the duplicate registration, you need to skip it when the function arguments are added https://github.com/omni-us/jsonargparse/blob/2de15ddfb1c02c2f7b3fe913ad11f13c5cb65dff/jsonargparse/_signatures.py#L166 https://github.com/Lightning-AI/litgpt/blob/main/litgpt/__main__.py#L121

And call `instantiate_class` here https://github.com/Lightning-AI/litgpt/blob/36c6a77435d75872f525848ee1570467d120ae80/litgpt/finetune/lora.py#L185-L187

This should be enough to unblock you. The not-so-nice thing is that the CLI args structure leaks into the actual script, meaning that users who don't go through the CLI will have to create this dictionary manually.

rasbt commented:
Awesome, thanks so much, this was great help! Figured it out now and got it to work. Many thanks, again learned something new!

rasbt commented:
I know got it to work as follows:

```
litgpt finetune full \
  --checkpoint_dir checkpoints/EleutherAI/pythia-160m 

# Specify optimizer and optimizer args:
litgpt finetune full \
  --checkpoint_dir checkpoints/EleutherAI/pythia-160m \
  --optimizer  torch.optim.SGD \
  --optimizer.init_args.lr 1000
```

But I feel like the way I am passing the optimizer kwargs seems a bit hacky. Is this there a built-in/better way to handle it @carmocca ? The thing is that when I pass an `--optimizer` argument it also passes additional kwargs to the setup:

```python
kwargs = {
    'optimizer.class_path': 'torch.optim.SGD',
    'optimizer.init_args.dampening': 0.0,
    'optimizer.init_args.differentiable': False,
    'optimizer.init_args.foreach': None,
    'optimizer.init_args.lr': 0.001,
    'optimizer.init_args.maximize': False,
    'optimizer.init_args.momentum': 0.0,
    'optimizer.init_args.nesterov': False,
    'optimizer.init_args.weight_decay': 0.0
}
```

That's why I added the parsing into class_path and init_args:

```python
    optimizer_class_path = None
    optimizer_init_args = {}
    for key, value in list(kwargs.items()):
        if key.startswith("optimizer"):
            if "class_path" in key:
                optimizer_class_path = value
            elif "init_args" in key:
                init_arg_key = key.split(".")[-1]
                optimizer_init_args[init_arg_key] = value
            del kwargs[key]
```

Everything seems to work, but I wonder if there isn't a better way to do it?

carmocca commented:
@rasbt I pushed a commit with what I would suggest. The `str` code path could be improved if we want to expose arguments like the learning rate outside of the CLI, but that should be straightforward to implement.

Also fyi, you don't need to specify the `.init_args` substring through command line

rasbt commented:
The only caveat now is that the class path still needs to be specified. I.e., only specifying the learning rate doesn't work

```
litgpt finetune full  --optimizer.lr 200  --checkpoint_dir checkpoints/EleutherAI/pythia-160m

error: Parser key "optimizer":
  Not a valid subclass of Optimizer. Got value: NestedArg(key='lr', val='200')
  Subclass types expect one of:
  - a class path (str)
  - a dict with class_path entry
  - a dict without class_path but with init_args entry (class path given previously)
```

And the optimizer always needs to be specified explicitely

```bash
litgpt finetune full  --optimizer AdamW --optimizer.lr 200  --checkpoint_dir checkpoints/EleutherAI/pythia-160m
```

Do you know if that's a jsonargparse thing, @carmocca ? Because we already set a default value in the `setup` method I was thinking that this is a bit weird.

carmocca commented on config_hub/finetune/tiny-llama/qlora.yaml:
```diff
@@ -85,18 +85,6 @@ train:
   # Whether to tie the embedding weights with the language modeling head weights. (type: Optional[bool], default: null)
   tie_embeddings:
 
-  #   (type: float, default: 0.0003)
-  learning_rate: 0.0002
-
-  #   (type: float, default: 0.02)
-  weight_decay: 0.0
-
-  #   (type: float, default: 0.9)
-  beta1: 0.9
-
-  #   (type: float, default: 0.95)
-  beta2: 0.95
-
```
I guess we don't care about backwards compatibility since the other CLI changes related to the checkpoint dir need to be done

rasbt replied:
Yeah, I think full backwards compatibility would probably also add so much complexity and overhead. I think we need to wrap all the CLI changes as soon as possible, and then we hopefully never have to do a breaking change like this again.

carmocca commented on litgpt/finetune/adapter.py:
```diff
@@ -69,6 +72,7 @@ def setup(
         data: Data-related arguments. If not provided, the default is ``litgpt.data.Alpaca``.
         train: Training-related arguments. See ``litgpt.args.TrainArgs`` for details.
         eval: Evaluation-related arguments. See ``litgpt.args.EvalArgs`` for details.
+        optimizer: An optimizer such as torch.optim.AdamW.
```
To be applied to the other files too:

```suggestion
        optimizer: An optimizer name (such as "AdamW") or config.
```

carmocca commented on litgpt/finetune/adapter_v2.py:
```diff
@@ -146,14 +150,23 @@ def main(
     if isinstance(fabric.strategy.precision, BitsandbytesPrecision):
-        import bitsandbytes as bnb
+        if (isinstance(optimizer, str) and "AdamW" not in optimizer) or (isinstance(optimizer, dict) and "AdamW" not in optimizer.get("class_path", "")):
+            raise ValueError("The chosen quantization format only supports the AdamW optimizer.")
 
-        optimizer_cls = bnb.optim.PagedAdamW
+        import bitsandbytes as bnb
+        if isinstance(optimizer, str):
+            optimizer = bnb.optim.PagedAdamW(model.parameters())
+        else:
+            optim_args = get_argument_names(bnb.optim.PagedAdamW)
+            allowed_kwargs = {key: optimizer["init_args"][key] for key in optim_args & optimizer["init_args"].keys()}
+            optimizer = bnb.optim.PagedAdamW(model.parameters(), **allowed_kwargs)
```
I would create a function out of the bitsandbytes optimizer creation code path and another one for else content (L164-168) since this is going to be the same code everywhere

carmocca reviewed (commented):
Added AdamW as the CLI default so that you don't have to specify it

rasbt commented:
I hope this is ready now @carmocca

carmocca commented on tests/test_utils.py:
```diff
@@ -306,4 +310,46 @@ def test_init_out_dir(tmp_path):
+
+
+def test_instantiate_bnb_optimizer_with_invalid_str(model_parameters):
+    with pytest.raises(ValueError):
```
Always add the "match" message in case you eat up an undesired error

```suggestion
    with pytest.raises(ValueError, match="only supports the AdamW"):
```

carmocca reviewed (approved):
Very nice job!

carmocca commented:
The azure failure does look real:

```python
>       fit(fabric, devices, state, train_dataloader, val_dataloader, out_dir, tokenizer_dir, train, eval, optimizer)
E       TypeError: fit() takes 9 positional arguments but 10 were given

/__w/6/s/extensions/thunder/pretrain.py:229: TypeError
----------------------------- Captured stderr call -----------------------------
Missing logger folder: /tmp/pytest-of-root/pytest-0/test_pretrain0/out/logs/tensorboard
Seed set to 42
=========================== short test summary info ============================
FAILED tests/test_thunder_pretrain.py::test_pretrain - TypeError: fit() takes 9 positional arguments but 10 were given
```

rasbt commented:
It does. Let me investigate ...

rasbt commented:
Should be fixed for good now @carmocca . I can switch the link to the original tinystories now that you have seen the green checks haha 😆
